sqlwhat/test_exercise.py

```diff
@@ -21,23 +23,33 @@ def test_exercise(
     """
     """
 
-    state = State(
-        student_code=student_code,
-        solution_code=solution_code,
-        pre_exercise_code=pre_exercise_code,
-        student_conn=student_conn,
-        solution_conn=solution_conn,
-        student_result=student_result,
-        solution_result=solution_result,
-        reporter=Reporter(errors=error),
-        force_diagnose=force_diagnose,
-    )
-
-    SCT_CTX["Ex"].root_state = state
+    reporter = Reporter(errors=error)
 
     try:
-        exec(sct, SCT_CTX)
-    except TestFail as tf:
-        return tf.payload
+        state = State(
```

Comment:
Is it needed to put State creation in the try except block?

Reply:
In pythonwhat we raise exceptions about parsing there. Creating the state here allows these exceptions to be treated similar to others.